### DISCUSSION TITLE
F16 aileron and elevator oscilations at CAS 550kn+

### DISCUSSION THREAD

**Discussion by Synergy91** [JSBSim-Team/jsbsim#675] [Support requests]
Hi JSBSim community,

Here is my issue:
Using the f-16 model configuration - when going over 550 kn CAS, I experience aileron and elevator oscillations.

My environment:
Windows 10
JSBSim v1.1.12; f-16 config file revision 1.95
JSBSim built as a static lib and linked in an Unreal 5 project.

Not sure if this is a bug or a malpractice on my part, but I can't find any info about this, except for this thread on a JSBsim mailing list:
https://sourceforge.net/p/jsbsim/mailman/jsbsim-devel/thread/1393953.48.1476232800676.JavaMail.bclay%40Mariah/
The direction I'm exploring in is the definitions of the FCS, as per the discussion in the thread linked above, but just can't figure out the issue.

Omer S.

**Comment by gallonmate** [comment]
From the sounds of the discussion, the current version 1.95 is the "fix" that Erik attempted, which likely improved things but didn't solve the oscillations you are seeing. 

First, to isolate any UE plugin issues, I would verify your game framerate is locked and smooth during flight. I added a small update to the plugin the other week, which locks jsb sim to 120hz, but you still have to manually lock/cap your game rate, it can be anything but just needs to hold steady. Otherwise this introduces another variable into the flight model. The other thing is you can take a look at one of the latest FlightGear F16 models, and try to extract the file/info needed to work with pure jsb sim. (Since flightgear model files tend to have a lot of extra flightgear specific data added.)

**Reply by Synergy91** [reply]
Works! I simply updated the UE5 plugin with your new update, and locked the game's frame rate to a constant, and oscillations of any kind were gone.
So I take from this that the 120hz fixed delta time for the FDM is that critical...

Thank you!

**Comment by bcoconni** [comment]
@Synergy91 given the lack of elements to support the bug claim, this support request is moved to discussions.

**Comment by bcoconni** [comment]
> Here is my issue:
Using the f-16 model configuration - when going over 550 kn CAS, I experience aileron and elevator oscillations.

We need more details about the conditions in which you are observing this behavior: are you using the f-16 model from JSBSim ? Have you modified it ? Is the autopilot used ? If so, which autopilot feature is used (altitude-holding, velocity-holding, etc.). Atmospheric conditions ? Wind heading, wind velocity, ...

Ideally, we need to be able to reproduce the reported issue to provide some help so do not hesitate to give details.

**Comment by seanmcleod** [comment]
It doesn't look like the PID gains are gain-scheduled based on CAS etc. so I wouldn't be surprised that there are regions where the fixed gains don't work well.

```xml
   <pid name="fcs/roll-rate-pid">
     <trigger>fcs/aileron-pid-trigger</trigger>
     <input>fcs/roll-trim-error</input>
     <kp> 3.00000 </kp>
     <ki> 0.00050 </ki>
     <kd> -0.00125 </kd>
   </pid>
```

**Reply by agodemar** [reply]
@seanmcleod nor the normalized roll rate is defined according to the current instantaneous airspeed:

https://github.com/JSBSim-Team/jsbsim/blob/48ac3951ff13f1682c8f03ece6c1f179b08e63f6/aircraft/f16/f16.xml#L356-L360

The `velocities/p-aero-rad_sec` is scaled by a constant. Usually the nondimensional roll rate is defined as: $\hat{p} = pb/2V$

**Comment by Synergy91** [comment]
Thanks everybody for the replies! 
it seems that the problem was that the FDM was executed in 60 hz, with a bit of a variance on the delta times. Fixing the FDM updates on 120hz eliminated the oscillations.
